nexus/db-queries/src/db/queries/virtual_provisioning_collection_update.rs

```diff
@@ -239,11 +239,52 @@ WITH
             virtual_provisioning_resource.id = ").param().sql("
           LIMIT
             1
+        ) = 1
+          AS update
+    ),")
+                .bind::<sql_types::Uuid, _>(id)
+            },
+            UpdateKind::DeleteInstance { id, max_instance_gen, .. } => {
+                // The filter condition here ensures that the provisioning record is
+                // only deleted if the corresponding instance has a generation
+                // number less than the supplied `max_instance_gen`. This allows a
+                // caller that is about to apply an instance update that will stop
+                // the instance and that bears generation G to avoid deleting
+                // resources if the instance generation was already advanced to or
+                // past G.
+                //
+                // If the relevant instance ID is not in the database, then some
+                // other operation must have ensured the instance was previously
+                // stopped (because that's the only way it could have been deleted),
+                // and that operation should have cleaned up the resources already,
+                // in which case there's nothing to do here.
+                query.sql("
+  do_update
```

Comment:
I basically tried to move all the validation of the state_generation into this boolean `do_update?` subquery, then re-use that check in multiple spots where we can perform updates below.